sentry-core/benches/scope_benchmark.rs

```diff
@@ -0,0 +1,159 @@
+//! `client` feature. To test without it, one needs to comment the circular dependency
+//! before running the benchmark.
+
+#[cfg(feature = "client")]
+use std::sync::Arc;
+
+use criterion::{criterion_group, criterion_main, Criterion};
+use sentry::protocol::Breadcrumb;
+#[cfg(not(feature = "client"))]
+use sentry_core as sentry;
+
+/// Tests Scopes with Breadcrumbs
+///
+/// This uses the [`sentry::add_breadcrumb`] API in *callback mode*, which means
+/// it is essentially a noop when the current Hub is inactive.
+fn scope_with_breadcrumbs(capture: bool) {
+    for i in 0..50 {
+        sentry::add_breadcrumb(|| Breadcrumb {
+            message: Some(format!("Breadcrumb {}", i)),
+            ..Default::default()
+        });
+    }
+
+    if capture {
+        sentry::capture_message("capturing on outer scope", sentry::Level::Info);
+    }
+
+    sentry::with_scope(
+        |_| (),
+        || {
+            // 50 + 70 exceeds the default max_breadcrumbs of 100
+            for i in 0..70 {
+                sentry::add_breadcrumb(|| Breadcrumb {
+                    message: Some(format!("Breadcrumb {}", i)),
+                    ..Default::default()
+                });
+            }
+
+            if capture {
+                sentry::capture_message("capturing within a nested scope", sentry::Level::Info);
+            }
+        },
+    );
+
+    sentry::configure_scope(|scope| scope.clear());
+}
+
+/// Tests Scopes with Tags
+///
+/// This uses the [`sentry::Scope::set_tag`] function to define, and then overwrite/extend
+/// the set of tags.
+fn scope_with_tags(capture: bool) {
+    sentry::configure_scope(|scope| {
+        for i in 0..20 {
+            scope.set_tag(&format!("tag {}", i), format!("tag value {}", i));
+        }
+    });
+
+    if capture {
+        sentry::capture_message("capturing on outer scope", sentry::Level::Info);
+    }
+
+    sentry::with_scope(
+        |scope| {
+            for i in 10..30 {
+                // since this is a hashmap, we basically overwrite 10, and add 10 new tags
+                scope.set_tag(&format!("tag {}", i), format!("tag value {}", i));
+            }
+        },
+        || {
+            if capture {
+                sentry::capture_message("capturing within a nested scope", sentry::Level::Info);
+            }
+        },
+    );
+
+    sentry::configure_scope(|scope| scope.clear());
+}
+
+/// Returns a new *active* [`sentry::Hub`] which discards Events in the Transport.
+#[cfg(feature = "client")]
+fn discarding_hub() -> sentry::Hub {
+    struct NoopTransport;
+
+    impl sentry::Transport for NoopTransport {
+        fn send_envelope(&self, envelope: sentry::Envelope) {
+            drop(envelope)
+        }
+    }
+
+    let client = Arc::new(sentry::Client::from(sentry::ClientOptions {
+        dsn: Some("https://public@sentry.invalid/1".parse().unwrap()),
+        // lol, this double arcing -_-
```

Comment:
lol, how did that happen?

Reply:
The first is `impl TransportFactory for Arc<Transport>`, the second one is because clientoptions takes an `Arc<dyn TransportFactory>` explicitly :-(